rbx_types/src/terrain.rs

```diff
@@ -0,0 +1,548 @@
+
+impl fmt::Display for InvalidMaterialError {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "Invalid binary material value {:x?}", self.0)
+    }
+}
+
+pub trait TerrainSerializer {
+    fn encode(&self) -> Vec<u8>;
+    //fn decode(&self) -> Self;
+}
+
+/// A container for a voxel of terrain, used in the `Chunk` object.
+#[derive(Debug, Default, PartialEq, Clone, Copy)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Voxel {
+    solid_occupancy: f32,
+    water_occupancy: f32,
+    material: Material,
+}
+
+impl Voxel {
+    /// Constructs a new `Voxel` with a material and occupancy percentage.
+    /// Equivalent to data writeable from Roblox's `Terrain:WriteVoxels`.
+    /// Occupancy values are between `0.0` and `1.0`, as a percentage of the voxel.
+    pub fn new(material: Material, solid_occupancy: f32) -> Self {
+        let mut voxel = Self {
+            material,
+            ..Default::default()
+        };
+        voxel.set_occupancy(solid_occupancy, 0.0);
+
+        voxel
+    }
+
+    /// Constructs a new `Voxel` with a material, solid occupancy, and water
+    /// occupancy percentage. Takes advantage of the recently-released
+    /// Shorelines feature. Equivalent to data writeable from Roblox's `Terrain:WriteVoxelChannels`.
+    /// Occupancy values are between `0.0` and `1.0`, as a percentage of the voxel.
+    pub fn new_with_water(material: Material, solid_occupancy: f32, water_occupancy: f32) -> Self {
+        let mut voxel = Self {
+            material,
+            ..Default::default()
+        };
+        voxel.set_occupancy(solid_occupancy, water_occupancy);
+
+        voxel
+    }
+
+    fn get_encode_data(&self) -> (u8, u8) {
+        let solid_occupancy: u8 = (self.solid_occupancy * 255.0) as u8;
+        let water_occupancy: u8 = (self.water_occupancy * 255.0) as u8;
+        (solid_occupancy, water_occupancy)
+    }
+
+    fn encode_run_length(&self, count: u16) -> Vec<u8> {
+        assert!(
+            (1..=256).contains(&count),
+            "Invalid voxel RLE count: {}",
+            count
+        );
+
+        let (solid_occupancy, water_occupancy) = self.get_encode_data();
+        let mut flag = self.material as u8;
+        let mut to_write: Vec<u8> = vec![];
+
+        if solid_occupancy != 0xFF && solid_occupancy != 0x00 {
+            // Should we store the solid occupancy value?
+            flag |= 0b01000000;
+            to_write.push(solid_occupancy);
+        }
+        if count > 1 || water_occupancy != 0 {
+            // Should we store the count (amount of voxels this run length) value?
+            flag |= 0b10000000;
+            if water_occupancy == 0 {
+                to_write.push((count - 1) as u8);
+            } else {
+                to_write.push(0);
+                to_write.push(water_occupancy);
+            }
+        }
+        to_write.insert(0, flag);
```

Comment:
```suggestion
        to_write[0] = flag;
```

This avoids having to reshuffle the vector every time.